test/parallel_api/iterator/transform_output_iterator.pass.cpp

```diff
@@ -0,0 +1,156 @@
+
+        ::std::copy(::std::forward<ExecutionPolicy>(exec), first1, last1, first2);
+
+        host_vals.retrieve_data();
+        auto res_begin = host_vals.get();
+        
+        EXPECT_EQ_N(expected_values, res_begin, n, "Wrong result from copy with transform_output_iterator");
+    }
+}; // struct test_copy
+
+void test_simple_copy(size_t buffer_size)
+{
+    // 1. create buffers - usm to test copying to transform output iterator
+    sycl::queue q{};
+
+    using TestBaseData = test_base_data_usm<sycl::usm::alloc::shared, int>;
+    TestBaseData test_base_data(q, {{ buffer_size, 0 }, 
+                                    { buffer_size, 0 }});
+
+    // 2. create iterators over source buffer 
+    auto sycl_source_begin = test_base_data.get_start_from(UDTKind::eKeys);
+    auto sycl_result_begin = test_base_data.get_start_from(UDTKind::eVals);
+
+    // 3. run algorithms
+    auto transformation = [](int item) { return item + 1; };
+
+    int identity = 0;
+    ::std::fill_n(sycl_source_begin, buffer_size, identity);
+    auto tr_host_result_begin = oneapi::dpl::make_transform_output_iterator(sycl_result_begin, transformation); 
+
+    ::std::vector<int> expected_res(buffer_size, identity + 1);
+
+    test_copy<int> test(test_base_data);
+    TestUtils::invoke_on_all_hetero_policies<1>()(test, sycl_source_begin, sycl_source_begin + buffer_size, 
+        tr_host_result_begin, buffer_size, expected_res.begin());
+}
+
+void test_multi_transform_copy(size_t buffer_size)
+{
+    // 1. create buffers
+    using TestBaseData = test_base_data_buffer<int>;
+    TestBaseData test_base_data({ { buffer_size, 0 },
+                                  { buffer_size, 0 } });
+
+    // 2. create iterators over buffers
+    sycl::buffer<int>& source_buf = test_base_data.get_buffer(UDTKind::eKeys);
+    sycl::buffer<int>& result_buf = test_base_data.get_buffer(UDTKind::eVals);
+
+    auto host_source_acc = source_buf.template get_access<sycl::access::mode::write>();
+    auto host_source_begin = host_source_acc.get_pointer();
+    auto sycl_result_begin = oneapi::dpl::begin(result_buf);
+
+    // 3. Run the algorithm
+
+    auto transformation1 = [](int item) { return item + 1; };
+    auto transformation2 = [](int item) { return item + 2; };
+    auto transformation3 = [](int item) { return item + 3; };
+
+    auto tr_host_source_begin = oneapi::dpl::make_transform_output_iterator(host_source_begin, transformation1);
+    auto tr2_host_source_begin = oneapi::dpl::make_transform_output_iterator(tr_host_source_begin, transformation2);
+    auto tr3_host_source_begin = oneapi::dpl::make_transform_output_iterator(tr2_host_source_begin, transformation3);
+    auto tr3_host_source_end = tr3_host_source_begin + buffer_size;
+
+    int identity = 0;
+    ::std::fill_n(tr3_host_source_begin, buffer_size, identity);
+    test_copy<int> test(test_base_data);
+
+    ::std::vector<int> expected_res(buffer_size, identity + 3);
```

Comment:
Is this actually the expected behavior of cascaded transform_output_iterators?
It seems like this only applies the last transformation, instead of cascading them on each other from last to first.  
Thrust's `transform_output_iterator` applies all transformations from outermost to innermost in order.  

I'm looking at why this is and how to resolve it...

Reply:
My most recent commit uses SFINAE to detect and collapse cascading `transform_output_ref_wrapper` objects in such a way that if you create a `transform_output_ref_wrapper` with a `transform_output_ref_wrapper` as its reference type, it will nest the unary functions, but keep the original reference from the inner `transform_output_ref_wrapper`.  When written to, all unary functions will be applied from outermost `transform_output_iterator` to the innermost in that order.

Also, since we only end up with a single layer of `transform_output_ref_wrapper`, we can make use of the conversion operator to the reference type.  Without collapsing the nested operators, this wasn't possible.

Reply:
These changes require C++14, for the use of auto return types without decltype helpers.  
With the composite lambda functions, return types become quite challenging here, and its non trivial to resolve this with decltype helpers it seems.  Since we are dropping C++11 support, is this something I should try to address?